pkg/branchcuts/bumper/replace.go

```diff
@@ -0,0 +1,78 @@
+package bumper
+
+import (
+	"fmt"
+	"regexp"
+	"sort"
+	"strconv"
+	"strings"
+)
+
+func ReplaceWithNextVersionInPlace(line *string, major int) error {
+	newLine, err := ReplaceWithNextVersion(*line, major)
+	if err != nil {
+		return err
+	}
+	*line = newLine
+	return nil
+}
+
+// Find every {major}.{minor} reference into 'line' and the replace it
+// with {major}.{minor+1}
+func ReplaceWithNextVersion(line string, major int) (string, error) {
+	p := fmt.Sprintf(`%d\.(?P<minor>\d+)`, major)
+	r := regexp.MustCompile(p)
+	m := r.FindAllStringSubmatch(line, -1)
+	if m == nil {
+		return line, nil
+	}
+
+	minors, err := uniqueSortedMinors(m)
+	if err != nil {
+		return line, err
+	}
+
+	for _, minor := range minors {
+		curVersion := fmt.Sprintf("%d.%d", major, minor)
+		nextVersion := fmt.Sprintf("%d.%d", major, minor+1)
+		line = strings.ReplaceAll(line, curVersion, nextVersion)
+	}
+
+	return line, nil
+}
+
+// The function extracts all the minors from matches, removes duplicate and finally it returs them in
+// decreasing order.
+//
+// Matches is an array of regex matches that could be obtained by the following example:
+// string: ocp_4.5-4.6-4.5
+// pattern: 4\.\d+
+// matches: [3]string{}
+// matches[0]: []string{ "4.5", "5" }
+// matches[1]: []string{ "4.6", "6" }
+// matches[2]: []string{ "4.5", "5" }
+//
+// Given the previous input, the function returns []int{6, 5}
+func uniqueSortedMinors(matches [][]string) ([]int, error) {
```

Comment:
This looks really smelly and adds a lot of complexity for no reason. Are you sure that this is the only way of doing this?

Reply:
When it comes to bump filenames we have to handle some cases like this:
`openshift-release-master__stable-4.7-upgrade-from-stable-4.6-from-stable-4.5.yaml`
that should be transformed into:
`openshift-release-master__stable-4.8-upgrade-from-stable-4.7-from-stable-4.6.yaml`.

In order to cope with that properly you should start from the biggest minor way to the smallest and keep bumping each of them (bottom-up simply doesn't work). So you catch all the versions `{"4.6","4.7","4.8",}` then  all the minors out of them `{"6","7","8"}` and finally in a descending order `{"8","7","6"}`. The routine `uniqueSortedMinors` tries to accomplish that. By the way I'm going to use the `sets` the reduce the lines of code if we stick with this approach.

That's not the only way, we could start from the minor of the version we're targeting, that is the future release, and repeat the processed I've described before. So if we're bumping to `4.13` (future release) the code would look like as follow:
```golang
// As an example: 
// line := "openshift-release-master__stable-4.12-upgrade-from-stable-4.11-from-stable-4.10.yaml"
major, minor := 4, 13
for minor := minor-1; minor >= 0; minor-- {
    curVersion := fmt.Sprintf("%d.%d", major, minor)
    nextVersion := fmt.Sprintf("%d.%d", major, minor+1)
    line = strings.ReplaceAll(line, curVersion, nextVersion)
}
// Result: 
// line := "openshift-release-master__stable-4.13-upgrade-from-stable-4.12-from-stable-4.11.yaml"

```
 That is going to work too and it will cycle a little bit more, but to be honest I don't think it doesn't make (and it won't make in the future too unless numbers become huge) any difference in terms of performance here.

Reply:
Why we can't just increment any `4.x` value?

Reply:
@droslean I don't think incrementing any `4.x` is a correct behavior if we are doing a branch cut for eg 4.12. Maybe it covers most of the cases, but not sure, this could lead to bumping something that is hard coded for example. That being said, it is true that normally we operate in ranges `4.x-2 `to `4.x+2`, and bumping anything in that range I would consider safe.

Reply:
We can bump any `4.x` value but the order the increment is done matters.

Consider again this line:
`openshift-release-master__stable-4.12-upgrade-from-stable-4.11-from-stable-4.10.yaml`
that has to be transformed to:
`openshift-release-master__stable-4.13-upgrade-from-stable-4.12-from-stable-4.11.yaml`

If we bump it in an order different from descending, the following scenario could happen:
1. Bumping from `4.10` to `4.11`
`openshift-release-master__stable-4.12-upgrade-from-stable-4.11-from-stable-4.11.yaml`
2. Bumping from `4.11` to `4.12`
`openshift-release-master__stable-4.12-upgrade-from-stable-4.12-from-stable-4.12.yaml`
3. Bumping from `4.12` to `4.13`
`openshift-release-master__stable-4.13-upgrade-from-stable-4.13-from-stable-4.13.yaml`

That's not what we expect.

Reply:
Bumping from `4.10` to `4.11`:
`openshift-release-master__stable-($TO+0.1)-upgrade-from-stable-$TO-from-stable-$TO.yaml`

Reply:
Either way. I am still not convinced that this is the best way. Why do we need to sort them? And what is this 2d string array? Ultimately, we just need to match a regex and increment it based on the given `FROM` and `TO` versions. This applies to everything.

Reply:
> Bumping from 4.10 to 4.11:
> `openshift-release-master__stable-($TO+0.1)-upgrade-from-stable-$TO-from-stable-$TO.yaml`

If what you intended was `TO="4.11"` then it will result in: 
`openshift-release-master__stable-4.12-upgrade-from-stable-4.11-from-stable-4.11.yaml`
and I'm not sure it's what we want.

If the starting string is:
`openshift-release-master__stable-4.12-upgrade-from-stable-4.11-from-stable-4.10.yaml`

the pattern it follows is:
`N=12`
`openshift-release-master__stable-4.$(( N ))-upgrade-from-stable-4.$(( N-1 ))-from-stable-4.$(( N-2 )).yaml`

therefore it should be transformed to (every minor version incremented by 1):
`openshift-release-master__stable-4.$(( N+1 ))-upgrade-from-stable-4.$(( N ))-from-stable-4.$(( N-1 )).yaml`

that is:
`openshift-release-master__stable-4.13-upgrade-from-stable-4.12-from-stable-4.11.yaml`

> Why do we need to sort them?

According to my approach we need to sort them because of [this](https://github.com/openshift/ci-tools/pull/2866#discussion_r1033431826).

> And what is this 2d string array?

It is the result of function [Regexp.FindAllStringSubmatch](https://pkg.go.dev/regexp#Regexp.FindAllStringSubmatch) from the standard library.

> I am still not convinced that this is the best way.

Feel free to suggest any code so we can discuss on it together.

Reply:
The reason I do not like this function is not necessarily the method or the complexity but the input, I feel it could be better than `[][]string` which is described not in the code but in the comment. I am ok, to extract this discussion and upgrade of this mechanism to another PR or minor Jira issue, just to prevent this whole tool from rotting.